### PR TITLE
Create organization api endpoints for multi-org

### PR DESCRIPTION
**Pull request by tzjames**
### Features and Changes
Creates api endpoints for posting a new organization and listing current organizations
Adds userId to ApiRequest
Adds validator function to make sure the request is a super admin request

### Testing
With a key of a super admin:
Testing put:
```
curl -X POST http://localhost:3100/api/v1/organizations \
  -d 'name=My subsidiary 4' \
  -u secret_user_4qK8JJ1f4Ge137YG8Jws88Z4oCYJbCeAIgLUii7wQ:
```
See the return be the new org:
```
{
  "organization": {
    "id": "org_19exnt4zjlneskts5",
    "name": "My subsidiary 4",
    "url": "",
    "verifiedDomain": "",
    "ownerEmail": "tallnerd@gmail.com",
    "dateCreated": "2023-10-06T15:59:27.172Z"
  }
}
```

Testing get:
```
curl  http://localhost:3100/api/v1/organizations \
 -u secret_user_4qK8JJ1f4Ge137YG8Jws88Z4oCYJbCeAIgLUii7wQ:
```
See the return be all the orgs:
```
  "organizations": [...]
  "limit": 10,
  "offset": 0,
  "count": 7,
  "total": 7,
  "hasMore": false,
  "nextOffset": null
```

Testing with limit:
```
curl  -G http://localhost:3100/api/v1/organizations \
 -d "limit=4" \
 -u secret_user_4qK8JJ1f4Ge137YG8Jws88Z4oCYJbCeAIgLUii7wQ:
```
See half the organizations:
```
  "organizations": [...]
  "limit": 4,
  "offset": 0,
  "count": 4,
  "total": 7,
  "hasMore": true,
  "nextOffset": 4
```

Testing with offset:
```
curl  -G http://localhost:3100/api/v1/organizations \
 -d "limit=4" \
 -d "offset=4" \
 -u secret_user_4qK8JJ1f4Ge137YG8Jws88Z4oCYJbCeAIgLUii7wQ:
```
See the remaining organizations:
```
  "limit": 4,
  "offset": 4,
  "count": 3,
  "total": 7,
  "hasMore": false,
  "nextOffset": null
```

Testing search:
```
curl  -G http://localhost:3100/api/v1/organizations \
 -d "search=Universal" \
 -u secret_user_4qK8JJ1f4Ge137YG8Jws88Z4oCYJbCeAIgLUii7wQ:
```
See the matching organizations:
```
{
  "organizations": [
    {
      "id": "org_19exnti39lnbwyae8",
      "name": "Universal Paperclips",
      "url": "",
      "verifiedDomain": "",
      "ownerEmail": "tallnerd@gmail.com",
      "dateCreated": "2023-10-04T15:38:35.167Z"
    }
  ],
  "limit": 10,
  "offset": 0,
  "count": 1,
  "total": 1,
  "hasMore": false,
  "nextOffset": null
}
```

With an API key:
```
curl http://localhost:3100/api/v1/organizations \              
  -u secret_admin_AfphevvY7p3C0vMgvsx16prAIcooJKQsvoTZWILm2c:
{
  "message": "This endpoint requires the use of a Personal Access Token rather than an API_KEY."
}
```

With a key from a non-superadmin user:

```
curl http://localhost:3100/api/v1/organizations \              
  -u secret_user_7gmXD8zBnZNkNVvhcvL1O0YDqxGfqihWVuZd6eA:
{
  "message": "This endpoint requires the Personal Access Token of a super admin."
}
```

### CODE REVIEW

File: packages/back-end/src/api/openapi/schemas/Organization.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  dateCreated:
+    type: string
+    format: date-time
+    description: The date the organization was created
+  verifiedDomain:
+    type: string
+    description: The verified domain for the organization
+  url:
+    type: string
+    description: The URL for the organization
```
Comment (jdorn) [inline packages/back-end/src/api/openapi/schemas/Organization.yaml:15]: I think we can remove `verifiedDomain` and `url` from here.

File: packages/back-end/src/api/organizations/postOrganization.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    let verifiedDomain = "";
+
+    // if the owner is verified, try to infer a verified domain
+    if (user.email && user.verified) {
+      const domain = user.email.toLowerCase().split("@")[1] || "";
+      const isFreeDomain = freeEmailDomains.includes(domain);
+      if (!isFreeDomain) {
+        verifiedDomain = domain;
+      }
+    }
```
Comment (jdorn) [inline packages/back-end/src/api/organizations/postOrganization.ts:34]: I don't think we want to set verified domains as part of this REST endpoint.  That's more for a user registering an organization for themselves

File: packages/back-end/src/util/handler.ts [hunk 3]
```diff
@@ -147,6 +149,54 @@ export function getBuild() {
   return build;
 }
 
+export async function validateIsSuperUserRequest(req: { userId?: string }) {
+  if (!IS_MULTI_ORG) {
+    throw new Error("This endpoint requires multi-org mode.");
+  }
```
Comment (jdorn) [inline packages/back-end/src/util/handler.ts:155]: Should also check that they have the enterprise feature for multiOrg.  Adnan just merged code today to enable this in REST endpoint, so if you pull in the latest main, it should be easy to add here.

Comment (tzjames) [reply packages/back-end/src/util/handler.ts:160]: So I added: ```  if (req.organization) {
    if (!orgHasPremiumFeature(req.organization, "multi-org")) {
      throw new Error("This endpoint requires an Enterprise plan.");
    }
  }``` Is that what you meant?

**Review by jdorn — APPROVED** [review]

**Pull request merged**